From js any

OlivierBlanvillain opened the pull request:
This PR adds a `def fromJsAny(json: js.Any): JValue` method for both safe and unsafe ATDs. 

As illustrated in the tests, the intent is to have the following law:

```
forAll { j: JValue => fromJsAny(j.toJsAny) == j }
```

I could not make this pass (I think) due to rounding errors on `JNumber`, depending on your feedback I could rework this with either static test cases or a solution to the rounding stuff.

OlivierBlanvillain committed:
Update testEquals with .copy()

OlivierBlanvillain committed:
Add JValue.fromJsAny

OlivierBlanvillain committed:
Add unsafe.JValue.fromJsAny

OlivierBlanvillain committed:
Document and @JSExport fromJsAny

mdedetrich commented on js/src/main/scala/scala/json/ast/JValue.scala:
```diff
@@ -31,6 +31,34 @@ sealed abstract class JValue extends Product with Serializable {
+@JSExport
+object JValue {
+  /**
+    * Converts a Javascript object/value coming from Javascript to a [[JValue]].
```
Can you say Scala.js Javascript object/value rather than just Javascript object/value. This is to clear confusion from Scala.js Javascript versus Javascript alone (which the AST also supports).

Also applies to anywhere else where it happens to be mentioned

OlivierBlanvillain replied:
I revered the doc for [toJsAny](https://github.com/mdedetrich/scala-json-ast/pull/7/files#diff-bbea03361f116097645be6b8f3000a28R31), but I'm not sure I understand the nuance here, isn't it the case that `js.Any` correspond to both Javascript values and Scala.js Javascript values?

mdedetrich commented on js/src/main/scala/scala/json/ast/JValue.scala:
```diff
@@ -31,6 +31,34 @@ sealed abstract class JValue extends Product with Serializable {
+    * @return
+    */
+  @JSExport
+  def fromJsAny(json: js.Any): JValue =
```
Shouldn't this have a return type of `Option[JValue]`? Not every `js.Any` can be converted to a `JValue`.

OlivierBlanvillain replied:
Makes sense :+1:

mdedetrich commented:
Thanks for the pull request.

I am not sure that it makes sense that any possible Scala.js javascript value can be converted to a JSON value. Doesn't `js.Any` also encapsulate types such as Javascript function and whatnot (which makes no sense in JSON whatsoever)? Sounds to me that it makes more sense for `fromJsAny` to be return an `Option[JValue]` rather than just a `JValue`.

I will look at the rounding error later today, definitely don't want an implementation that is needlessly losing numeric data but I have to see how Scala.js handles numbers in more detail (as to why this happening).

OlivierBlanvillain committed:
Merge branch 'master' into from-js-any

OlivierBlanvillain committed:
Rework fromJsAny to return an `Option[JValue]`

For reference, this implementation was derived from https://goo.gl/4gmC13
(which return a precise error showing the parsing failed), which was in
turned inspired from https://goo.gl/iQ0ANV.

OlivierBlanvillain commented:
Sorry for taking so long (and for such little rework). BTW the problem on number rounding still remains; I did not investigated it at all...

mdedetrich commented:
Thanks, I will have a look into the rounding error

mdedetrich commented:
I am going to close this. @OlivierBlanvillain will open it when I you update your PR after https://github.com/mdedetrich/scalajson/pull/30 is done since a lot of changes have happened since then.
